src/lib/mod.rs

```diff
@@ -122,35 +163,59 @@ pub fn get_agent(pem: &str) -> AnyhowResult<Agent> {
 
+    let identity = get_identity(auth)?;
     builder
-        .with_boxed_identity(get_identity(pem))
+        .with_boxed_identity(identity)
         .build()
         .map_err(|err| anyhow!(err))
 }
 
-/// Returns an identity derived from the private key.
-pub fn get_identity(pem: &str) -> Box<dyn Identity + Sync + Send> {
-    if pem.is_empty() {
-        return Box::new(AnonymousIdentity);
-    }
-    match Secp256k1Identity::from_pem(pem.as_bytes()) {
-        Ok(identity) => Box::new(identity),
-        Err(_) => match BasicIdentity::from_pem(pem.as_bytes()) {
-            Ok(identity) => Box::new(identity),
-            Err(_) => {
-                eprintln!("Couldn't load identity from PEM file");
-                std::process::exit(1);
-            }
-        },
+fn ask_nitrohsm_pin_via_tty() -> Result<String, String> {
+    rpassword::read_password_from_tty(Some("NitroHSM PIN: "))
+        .context("Cannot read NitroHSM PIN from tty")
+        // TODO: better error string
+        .map_err(|e| e.to_string())
+}
+
+fn read_nitrohsm_pin_env_var() -> Result<Option<String>, String> {
+    match std::env::var("NITROHSM_PIN") {
+        Ok(val) => Ok(Some(val)),
+        Err(VarError::NotPresent) => Ok(None),
+        Err(e) => Err(format!("{}", e)),
     }
 }
 
-pub fn require_pem(pem: &Option<String>) -> AnyhowResult<String> {
-    match pem {
-        None => Err(anyhow!(
-            "Cannot use anonymous principal, did you forget --pem-file <pem-file> ?"
-        )),
-        Some(val) => Ok(val.clone()),
+/// Returns an identity derived from the private key.
+pub fn get_identity(auth: &AuthInfo) -> AnyhowResult<Box<dyn Identity>> {
+    match auth {
+        AuthInfo::NoAuth => Ok(Box::new(AnonymousIdentity) as _),
+        AuthInfo::PemFile(pem) => match Secp256k1Identity::from_pem(pem.as_bytes()) {
+            Ok(id) => Ok(Box::new(id) as _),
+            Err(_) => match BasicIdentity::from_pem(pem.as_bytes()) {
+                Ok(id) => Ok(Box::new(id) as _),
+                Err(e) => Err(e).context("couldn't load identity from PEM file"),
+            },
+        },
+        AuthInfo::NitroHsm(info) => {
+            let pin_fn = || {
+                let user_set_pin = { info.pin.borrow().clone() };
+                match user_set_pin {
+                    None => match read_nitrohsm_pin_env_var() {
+                        Ok(Some(pin)) => Ok(pin),
+                        Ok(None) => {
+                            let pin = ask_nitrohsm_pin_via_tty()?;
+                            *info.pin.borrow_mut() = Some(pin.clone());
+                            Ok(pin)
+                        }
+                        Err(e) => Err(e),
+                    },
+                    Some(pin) => Ok(pin),
+                }
+            };
+            HardwareIdentity::new(&info.libpath, info.slot, &info.ident, pin_fn)
+                .context("Unable to use your hardware key")
+                .map(|i| Box::new(i) as _)
```

Comment:
Ah, I see, I think this way will be more readable:

```suggestion
            Box::new(HardwareIdentity::new(&info.libpath, info.slot, &info.ident, pin_fn)
                .context("Unable to use your hardware key")?)
```